src/util/alloc/malloc_allocator.rs

```diff
@@ -35,15 +36,21 @@ impl<VM: VMBinding> Allocator<VM> for MallocAllocator<VM> {
     fn alloc_slow_once(&mut self, size: usize, align: usize, offset: isize) -> Address {
         assert!(offset >= 0);
 
-        let ret = self.space.alloc(self.tls, size, align, offset);
-        trace!(
-            "MallocSpace.alloc size = {}, align = {}, offset = {}, res = {}",
-            size,
-            align,
-            offset,
-            ret
-        );
-        ret
+        loop {
+            let ret = self.space.alloc(self.tls, size, align, offset);
+            if object_ref_may_cross_chunk::<VM>(ret) {
+                self.space.free(ret);
```

Comment:
If the malloc implementation is clever enough to reuse the memory you just `free`-ed in the next `malloc`, this loop may go on forever.  We need to think about a way to "disrupt" `malloc` a bit, for example, allocating a slightly bigger object, requiring a stricter alignment, do the second `malloc` before `free`-ing, etc.

Reply:
I made changes to the code. If the first allocation does not pass the object ref check, a vec is created to store the 'bad' results, and the results are free'd when we have a 'good' result.

Reply:
Kunshan’s concern is very valid. For example, the malloc in glibc has tcache, which can be problematic.

Yi’s new approach sounds better. But it still has one potential problem. For example, if you allocate A and it’s bad, so you allocate again, you get B, which is good, and then you free A. A goes into the tcache, and for next allocation of the same size class, you will get A again, which you need to withhold, and try again. That is, once you get a bad cell and free it, it stays forever in the tcache (and it can grow the next time you hit a chunk boundary). I’m not sure whether this will manifest in real world scenarios, or whether this will introduce a performance hit (we assume that if people call into libraries for malloc, they probably don’t care too much about performance anyway)